### PR TITLE
[ci] Add gen_snapshot artifacts

### PR DESCRIPTION
**Pull request by WonyoungChoi**
- #224 

Add jobs for building gen_snapshot on windows / macOS.

Added artifacts:
- tizen-arm-release_linux-x64
- tizen-arm-release_windows-x64
- tizen-arm-release_darwin-x64
- tizen-arm-profile_linux-x64
- tizen-arm-profile_windows-x64
- tizen-arm-profile_darwin-x64
- tizen-arm64-release_linux-x64
- tizen-arm64-release_windows-x64
- tizen-arm64-release_darwin-x64
- tizen-arm64-profile_linux-x64
- tizen-arm64-profile_windows-x64
- tizen-arm64-profile_darwin-x64

### CODE REVIEW

**Comment by swift-kim** [comment]
```
TIZEN_ENGINE_GITHUB_RUN_ID=1614508761 flutter-tizen precache --tizen --force
```
```
$ flutter-tizen run --release
Launching tizen/flutter/generated_main.dart on Tizen SM-R500 in release mode...
ProcessException: Found candidates, but lacked sufficient permissions to execute
"/Users/swift/Git/flutter-tizen/flutter/bin/cache/artifacts/engine/tizen-arm-release/darwin-x64/gen_snapshot".
  Command: /Users/swift/Git/flutter-tizen/flutter/bin/cache/artifacts/engine/tizen-arm-release/darwin-x64/gen_snapshot
Building a Tizen application in release mode...                    349ms
Exception: The build failed.
```
Worked on my watch (M1 Macbook as a host) after providing the executable permission to the `gen_snapshot` binary. Tested both release and profile modes.

**Comment by WonyoungChoi** [comment]
😢 I will test on my Intel MacBook. but is it possible to run x64 executable on M1? What happens if you run the `/Users/swift/Git/flutter-tizen/flutter/bin/cache/artifacts/engine/tizen-arm-release/darwin-x64/gen_snapshot` executable directly?

**Comment by WonyoungChoi** [comment]
Successful on Intel MacBook
Result:
```
xamarin@Kanghos-MacBook-Air myapp1 % flutter-tizen run --release
Launching tizen/flutter/generated_main.dart on Tizen rpi4 in release mode...
The TPK was signed with a default certificate. You can create one using Certificate Manager.
https://github.com/flutter-tizen/flutter-tizen/blob/master/doc/install-tizen-sdk.md#create-a-tizen-certificate
Building a Tizen application in release mode...                    42.4s
✓ Built build/tizen/tpk/com.example.myapp1-1.0.0.tpk (6.6MB).
Installing build/tizen/tpk/com.example.myapp1-1.0.0.tpk...          4.7s

Flutter run key commands.
h List all available interactive commands.
c Clear the screen
q Quit (terminate the application on the device).
```

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -101,6 +101,122 @@ jobs:
...
+        run: |
+          gclient config --name=src\flutter --unmanaged https://github.com/flutter-tizen/engine
+          $env:PYTHONPATH="C:\workspace\depot_tools"
+          python3 src\flutter\ci\tizen\gclient-shallow-sync.py src\flutter\DEPS
+          gclient sync -v --no-history --shallow
+
+      - name: Build
+        working-directory: C:\workspace\engine\src
+        run: |
+          python3 .\flutter\tools\gn --target-os=linux --linux-cpu=${{ matrix.arch }} --no-goma --runtime-mode=${{ matrix.mode }}
```
Comment (swift-kim) [inline .github/workflows/build.yml:154]: For readability and consistency
```suggestion
          python3 .\flutter\tools\gn `
            --linux --linux-cpu=${{ matrix.arch }} `
            --no-goma `
            --runtime-mode=${{ matrix.mode }}
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -101,6 +101,122 @@ jobs:
...
+          ninja -C .\out\linux_${{ matrix.mode }}_${{ matrix.arch }} gen_snapshot
+
+      - uses: actions/upload-artifact@v2
+        with:
+          name: tizen-${{ matrix.arch }}-${{ matrix.mode }}_windows-x64
+          path: |
+            C:\workspace\engine\src\out\linux_${{ matrix.mode }}_${{ matrix.arch }}\gen_snapshot.exe
```
Comment (swift-kim) [inline .github/workflows/build.yml:161]: The line break doesn't seem necessary.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -101,6 +101,122 @@ jobs:
...
+          echo "$PWD/depot_tools" >> $GITHUB_PATH
+      
+      - name: Install required packages
+        run: |
+          brew install md5sha1sum
+
+      - name: GClient Sync
+        run: |
+          gclient config --name=src/flutter --unmanaged https://github.com/flutter-tizen/engine
+          gclient setdep --var=download_android_deps=False --deps-file=src/flutter/DEPS
```
Comment (swift-kim) [inline .github/workflows/build.yml:198]: Is this line only required for macos build?

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -101,6 +101,122 @@ jobs:
...
+          gclient sync -v --no-history --shallow
+      
+      - name: Build
+        run: |
+          src/flutter/ci/tizen/cache-checksum.sh restore src/out/$OUTPUT_NAME
+          # Change host_toolchain to mac/clang_x64.
+          sed -i "" "s|//build/toolchain/linux:clang_$host_cpu|//build/toolchain/mac:clang_$host_cpu|g" src/build/config/BUILDCONFIG.gn
+          # Pass dummy toolchain values to use custom toolchain settings.
+          python3 ./src/flutter/tools/gn --linux --linux-cpu=${{ matrix.arch }} --runtime-mode=${{ matrix.mode }} \
+            --no-goma --macos-build-tizen \
```
Comment (swift-kim) [inline .github/workflows/build.yml:209]: For consistency
```suggestion
          python3 ./src/flutter/tools/gn \
            --linux --linux-cpu=${{ matrix.arch }} \
            --runtime-mode=${{ matrix.mode }} \
            --no-goma \
```
By the way haven't we agreed to remove the `--macos-build-tizen` option?

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -101,6 +101,122 @@ jobs:
...
+            --target-toolchain _ --target-sysroot _ --target-triple _
+          ninja -C ./src/out/$OUTPUT_NAME clang_x64/gen_snapshot
+          src/flutter/ci/tizen/cache-checksum.sh save src/out/$OUTPUT_NAME
+
+      - uses: actions/upload-artifact@v2
+        with:
+          name: tizen-${{ matrix.arch }}-${{ matrix.mode }}_darwin-x64
+          path: |
+            src/out/linux_${{ matrix.mode }}_${{ matrix.arch }}/clang_x64/gen_snapshot
```
Comment (swift-kim) [inline .github/workflows/build.yml:218]: Same here.
```suggestion
          path: src/out/linux_${{ matrix.mode }}_${{ matrix.arch }}/clang_x64/gen_snapshot
```

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -101,6 +101,122 @@ jobs:
...
+          echo "GYP_MSVS_OVERRIDE_PATH=C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise" >> $Env:GITHUB_ENV
+          echo "WINDOWSSDKDIR=C:\Program Files (x86)\Windows Kits\10" >> $Env:GITHUB_ENV
+
+      - name: Install depot_tools
+        run: |
+          Invoke-WebRequest -Uri https://storage.googleapis.com/chrome-infra/depot_tools.zip -OutFile depot_tools.zip
+          7z x -y -o"C:\workspace\depot_tools" .\depot_tools.zip
+          echo "C:\workspace\depot_tools" >> $Env:GITHUB_PATH
+
+      - name: GClient Sync
```
Comment (swift-kim) [inline .github/workflows/build.yml:143]: Why don't you decapitalize the letters for consistency with the Linux step name?

Comment (WonyoungChoi) [reply .github/workflows/build.yml:143]: ok, I will change all step names to lowercase letters.

File: .github/workflows/build.yml (hunk 3 above)
Comment (WonyoungChoi) [reply .github/workflows/build.yml:198]: In Windows, the default value of `download_android_deps` is False.

File: .github/workflows/build.yml (hunk 4 above)
Comment (WonyoungChoi) [reply .github/workflows/build.yml:209]: I will remove `--macos-build-tizen` parameter and fix the `gn` to use the required settings.

**Review by swift-kim — APPROVED** [review]
Thank you!

**Pull request merged**